kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/NettyFilterDispatchExecutorTest.java

```diff
@@ -111,7 +126,68 @@ void completeOnFilterDispatchThreadWhenInEventLoop() {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    static <T> CompletableFuture<T> toStandardCompletableFuture(CompletionStage<T> stage) {
+        CompletableFuture<T> future = new CompletableFuture<>();
+        stage.whenComplete((o, throwable) -> {
+            if (throwable != null) {
+                future.completeExceptionally(throwable);
+            }
+            else {
+                future.complete(o);
+            }
+        });
+        return future;
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadResultCannotBeConvertedToCompletableFuture() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Object> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage);
+            assertThatThrownBy(inDispatchThread::toCompletableFuture)
+                    .isInstanceOf(UnsupportedOperationException.class)
+                    .hasMessage("CompletableFuture usage disallowed, we don't want to block the event loop or allow unexpected completion");
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadResultNotCompletableByClientByCasting() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Object> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage);
+            boolean isCompletableFuture = inDispatchThread instanceof CompletableFuture;
+            // cannot assertThat(inDispatchThread) because it attempts to use toCompletableFuture immediately
+            assertThat(isCompletableFuture).isFalse();
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadResultChainedAsyncOperationsDefaultToDispatchThread() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Object> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage);
+            CompletionStage<Boolean> chainedAsyncStage = inDispatchThread.thenApplyAsync(o -> dispatchExecutor.isInFilterDispatchThread());
+            completionStage.complete("abc");
+            assertThat(toStandardCompletableFuture(chainedAsyncStage)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
```

Comment:
I get a yellow warning in the IDE on the isEqualTo.. if you pass an IOAF constant, you can clear it.  You've got a few instances like this. WDYT?

```suggestion
            assertThat(toStandardCompletableFuture(chainedAsyncStage)).succeedsWithin(5, TimeUnit.SECONDS, InstanceOfAssertFactories.BOOLEAN).isEqualTo(true);
```